st2common/st2common/transport/connection_retry_wrapper.py

```diff
@@ -0,0 +1,164 @@
+# Licensed to the StackStorm, Inc ('StackStorm') under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import eventlet
+
+__all__ = ['ConnectionRetryWrapper', 'ClusterRetryContext']
+
+
+class ClusterRetryContext(object):
+    """
+    Stores retry context for cluster retries. It makes certain assumptions
+    on how cluster_size and retry should be determined.
+    """
+    def __init__(self, cluster_size):
+        # No of nodes in a cluster
+        self.cluster_size = cluster_size
+        # No of times to retry in a cluster
+        self.cluster_retry = 2
+        # time to wait between retry in a cluster
+        self.wait_between_cluster = 10
+
+        # No of nodes attempted. Starts at 1 since the
+        self._nodes_attempted = 1
+
+    def test_should_stop(self):
+        should_stop = True
+        if self._nodes_attempted > self.cluster_size * self.cluster_retry:
+            return should_stop, -1
+        wait = 0
+        should_stop = False
+        if self._nodes_attempted % self.cluster_size == 0:
+            wait = self.wait_between_cluster
+        self._nodes_attempted += 1
+        return should_stop, wait
+
+
+class ConnectionRetryWrapper(object):
```

Comment:
Yes I wanted to believe so too but guess what those are all lies that kombu is spreading!

So there are 4 error see https://github.com/celery/kombu/blob/master/kombu/connection.py#L781-L811.

Of these there are some which are characterized as recoverable errors. 

If you look at the source of http://kombu.readthedocs.org/en/latest/reference/kombu.html#kombu.Connection.autoretry and consequently http://kombu.readthedocs.org/en/latest/reference/kombu.html#kombu.Connection.ensure you will see that it only handles the recoverable errors i.e `recoverable_connection_errors` and `recoverable_channel_errors`. 

We already use that function for dealing with the recoverable errors (thank to @lakshmi-kannan) and we still have the code https://github.com/StackStorm/st2/pull/1760/files#diff-cd01bec138927aee7af667f279ff7358R163.

For the "unrecoverable" errors which are thrown when a node in a cluster goes down we are on the hook to deal with the error. `ConsumerMixin` actually deals with this part https://github.com/celery/kombu/blob/master/kombu/mixins.py#L172 but I could not find any Producer code which handled this in any reasonable manner.

If you see the impl https://github.com/StackStorm/st2/pull/1760/files#diff-cd01bec138927aee7af667f279ff7358R113 you will see that net is resembles the `ConsumerMixin` code which is spread across multiple methods.

## Some wishes -
* I wish I could use some kombu primitives to handle node failures properly.
* I wish kombu APIs were properly documented and set exact expectations so I did not have to read their horrible and convoluted code. (its not that bad I was probably just annoyed)
* I wish someone updates kombu to write their producers nicely and deal with ConnectionPools etc like any self respecting remote service connection library should.
* I wish I can burn all this code I have written once some library handles this mess.